README
Redis Sharding
**************

https://github.com/kni/redis-sharding

Redis Sharding is a multiplexed proxy-server, designed to work with the database divided to several servers.
It's a temporary substitution of Redis Cluster (http://redis.io) that is under development.

Redis Sharding is used for horizontal Redis database scaling (with connecting of additional servers) as long as load distribution between the cores on the multiprocessor servers (as Redis server is single-threaded, several copies of the server can be run, one for each free core).

                              /- Redis (node 1)
 Client 1 ---                /-- Redis (node 2)
              Redis Sharding --- Redis (node 3)
 Client 2 ---                \-- Redis (node 4)
                              \- Redis (node 5)

Sharding is done based on the CRC32 checksum of a key or key tag ("key{key_tag}").

The majority of the commands are supported except the key renaming commands, 
commands of working with sets, transactions, subscription and, of course, server configure. 

For BLPOP and BRPOP commands all the keys should be on one node, that is done with the key tag.
Also for this commands Redis timeout 0 is recommended. 

Perl 5.10 or higher is needed with the following modules installed: String::CRC32, EV.

Launch:

 perl redis_sharding.pl                             --nodes=10.1.1.2:6380,10.1.1.3:6380,...
 perl redis_sharding.pl                 --port=6379 --nodes=10.1.1.2:6380,10.1.1.3:6380,...
 perl redis_sharding.pl --host=10.1.1.1 --port=6379 --nodes=10.1.1.2:6380,10.1.1.3:6380,...


Resharding
----------

Cluster configuration change on the fly is not supported. We are looking forward for Redis Cluster release. 

To copy data to a new cluster use a new utility resharding.pl.

For example, there is a cluster consisting of two servers: A1 and A2. It is served with redis_sharding, that is launched as

 perl redis_sharding.pl --nodes=A1,A2

We want to copy data from the base 9 to a new cluster of 5 servers: B1, B2, B3, B4 and B5.

To do this we need to stop redis_sharding.pl and run resharding.pl for each server of the cluster: 

 perl resharding.pl --db=9 --from=A1 --nodes=B1,B2,B3,B4,B5
 perl resharding.pl --db=9 --from=A2 --nodes=B1,B2,B3,B4,B5

After copying have been done, run redis_sharding.pl for a new cluster: 

perl redis_sharding.pl --nodes=B1,B2,B3,B4,B5

If a flag --flushdb is set, the FLUSHDB command is sent to all cluster nodes before data copying.


Changes.txt
2011-05-25	0.3
	fix error on responce ":i", where i < 0

2011-03-11	0.2
	the usage of RAM has been reduced

2011-01-25	0.1
	first production release


RedisSharding.pm

package RedisSharding;

use strict;
use warnings;

use String::CRC32;

use Exporter;
our @ISA = qw(Exporter);
our @EXPORT = qw(cmd2stream get_client_reader get_servers_reader readers);







sub get_client_reader {
	my ($sub_cmd, $DEBUG) = @_;

	my $buf           = "";
	my $i             = 0;
	my $args_cnt      = 0;
	my $size_next_arg = 0;
	my @args          = ();

	my $sub_clean = sub {
		$i             = 0;
		$args_cnt      = 0;
		$size_next_arg = 0;
		@args          = ();
	};

	return sub {
		my ($_buf) = @_;
		$buf .= $_buf;
		NEXT: {
			if (not $args_cnt) {
				$buf =~ s/^(\015\012)+//;
				return 1 unless $buf;
				if ((index $buf, '*', 0) == 0) {
					if ((my $_i = index $buf, "\015\012", 1) > -1) {
						$args_cnt = substr $buf, 1, $_i - 1;
						print "args_cnt: $args_cnt\n" if $DEBUG;
						$i = $_i + 2;
					}
				} else {
					print "PARSING ERROR: args_cnt\n" if $DEBUG;
					$buf = ""; $sub_clean->();
					return;
				}
			}
			return 1 unless length $buf > $i;
			if ($args_cnt and not $size_next_arg) {
				if ((index $buf, '$', $i) == $i) {
					if ((my $_i = index $buf, "\015\012", $i + 1) > -1) {
						$size_next_arg = substr $buf, $i + 1, $_i - ($i + 1);
						print "size_next_arg: $size_next_arg\n" if $DEBUG;
						$i = $_i + 2;
					}
				} else {
					print "PARSING ERROR: size_next_arg\n" if $DEBUG;
					$buf = ""; $sub_clean->();
					return;
				}
			}
			return 1 unless length $buf > $i;
			if ($args_cnt and $size_next_arg) {
				if ((index $buf, "\015\012", $i + $size_next_arg) == $i + $size_next_arg) {
					my $arg = substr $buf, $i, $size_next_arg;
					print "arg: $arg\n" if $DEBUG;
					push @args, $arg;
					$i = $i + $size_next_arg + 2;
					$size_next_arg = 0;
					if ($args_cnt == @args) {
						print "NEXT COMMAND\n" if $DEBUG;
						$sub_cmd->(@args);
						substr $buf, 0, $i, "";
						$sub_clean->();
					}
					redo NEXT if length $buf;
				} elsif (length $buf > $i + $size_next_arg + 2) {
					print "PARSING ERROR: arg\n" if $DEBUG;
					$buf = ""; $sub_clean->();
					return;
				}
			}
		}

		return 1;
	};
}






sub get_servers_reader {
	my %args = @_;

	my %buf = ();

	my ($cmd, @s);

	my %cmd = ();

	my $reply_type = "";

	my %bulk_args     = ();
	my %bulk_args_had = ();
	my %bulk_size     = ();
	my $size_all      = 0; 

	my $sub_clean = sub {
		($cmd, @s) = ();
		%cmd = ();
		$reply_type = "";
		%bulk_args     = ();
		%bulk_args_had = ();
		%bulk_size     = ();
		$size_all      = 0;
	};


	return sub {
		my ($s, $buf) = @_;
		$buf{$s} .= $buf;

		my $next;
		NEXT: {
			$next = 0;
			unless ($cmd) {
				($cmd, @s) = $args{sub_cmd}->();
				@s or @s = @{$args{servers}};
			}

			my $re_line   = qr/([-+].+?|:-?\d+)\015\012/;
			my $re_bulk_1 = qr/\*(\d+|-1)\015\012/;
			my $re_bulk_2 = qr/\$(\d+|-1)\015\012/;

			foreach my $s (@s) {
				unless ($cmd{$s}) {
					if ($buf{$s}) {
						if ($reply_type =~ m/^(|line)$/ and $buf{$s} =~ m/^$re_line/) {
							print "RESPONSE from $s on $cmd: $1\n" if $args{DEBUG};
							$reply_type or $args{sub_response_type}->("line");
							$reply_type ||= "line";
							$args{sub_line_response}->($s, $1);
							$cmd{$s} = $1;
							substr $buf{$s}, 0, $+[0], "";
							$next = 1;
						} elsif ($reply_type =~ m/^(|multi_bulk)$/ and not exists $bulk_args{$s} and $buf{$s} =~ m/^$re_bulk_1/) {
							print "RESPONSE from $s on $cmd: *$1\\r\\n\n" if $args{DEBUG};
							$reply_type or $args{sub_response_type}->("multi_bulk");
							$reply_type ||= "multi_bulk";
							if ($1 == -1) {
								$bulk_args{$s} = undef;
								$args{sub_bulk_response_size}->($s, undef);
								$bulk_args_had{$s}++;
								$cmd{$s} = 1;
							} elsif ($1 == 0) {
								$bulk_args{$s} = 0;
								$args{sub_bulk_response_size}->($s, 0);
								$bulk_args_had{$s}++;
								$cmd{$s} = 1;
							} else {
								$bulk_args{$s} = $1;
								$args{sub_bulk_response_size}->($s, $1);
							}
							substr $buf{$s}, 0, $+[0], "";
							$next = 1;
						} elsif ($reply_type =~ m/^(|bulk)$/ and not exists $bulk_args{$s} and $buf{$s} =~ m/^$re_bulk_2/) {
							print "RESPONSE from $s on $cmd: \$$1\\r\\n\n" if $args{DEBUG};
							$reply_type or $args{sub_response_type}->("bulk");
							$reply_type ||= "bulk";
							if ($1 == -1) {
								$bulk_args{$s} = undef;
								$args{sub_bulk_response_size}->($s, undef);
								$bulk_args_had{$s}++;
								$cmd{$s} = 1;
							} else {
								$bulk_args{$s} = 1;
								$bulk_size{$s} = $1;
								$args{sub_bulk_response_size}->($s, 1);
							}
							substr $buf{$s}, 0, $+[0], "";
							$next = 1;
						} elsif ($reply_type =~ m/^multi_bulk$/ and not exists $bulk_size{$s} and $buf{$s} =~ m/^$re_bulk_2/) {
							if ($1 == -1) {
								$args{sub_bulk_response_arg}->($s, undef);
								$bulk_args_had{$s}++;
								$cmd{$s} = 1 if $bulk_args{$s} == $bulk_args_had{$s};
							} else {
								$bulk_size{$s} = $1;
							}
							print "RESPONSE from $s on $cmd: \$", $bulk_size{$s} // "-1", "\\r\\n\n" if $args{DEBUG};
							substr $buf{$s}, 0, $+[0], "";
							$next = 1;
						} elsif ($reply_type =~ m/^(multi_)?bulk$/ and defined $bulk_size{$s}) {
							if ((index $buf{$s}, "\015\012", $bulk_size{$s} - 1) == $bulk_size{$s}) {
								my $buf = substr $buf{$s}, 0, $bulk_size{$s};
								print "RESPONSE from $s on $cmd: $buf\\r\\n\n" if $args{DEBUG};
								$args{sub_bulk_response_arg}->($s, $buf);
								$bulk_args_had{$s}++;
								$cmd{$s} = 1 if $bulk_args{$s} == $bulk_args_had{$s};
								substr $buf{$s}, 0, $bulk_size{$s} + 2, "";
								delete $bulk_size{$s};
								$next = 1;
							} elsif (length $buf{$s} >= $bulk_size{$s} + 2) {
								warn "PARSING ERROR for $cmd";
							}
						}
					}
				}
			}

			if (not $size_all and $reply_type =~ m/^(multi_)?bulk$/ and @s eq keys %bulk_args) {
				$args{sub_bulk_response_size_all}->();
				$size_all = 1;
			}
			if (@s eq keys %cmd) {
				print "RESPONSE from all on $cmd\n" if $args{DEBUG};
				$args{sub_response_received}->();
				$sub_clean->();
				$next = 1;
			}

			redo NEXT if $next;
		}
	};
}





my %cmd_type = ();

$cmd_type{$_} = 1 for qw(PING AUTH SELECT FLUSHDB FLUSHALL DBSIZE KEYS);

$cmd_type{$_} = 2 for qw(
EXISTS TYPE EXPIRE PERSIST TTL MOVE
SET GET GETSET SETNX SETEX
INCR INCRBY DECR DECRBY APPEND SUBSTR
RPUSH LPUSH LLEN LRANGE LTRIM LINDEX LSET LREM LPOP RPOP
SADD SREM SPOP SCARD SISMEMBER SMEMBERS SRANDMEMBER
ZADD ZREM ZINCRBY ZRANK ZREVRANK ZRANGE ZREVRANGE ZRANGEBYSCORE ZCOUNT ZCARD ZSCORE ZREMRANGEBYRANK ZREMRANGEBYSCORE
HSET HGET HMGET HMSET HINCRBY HEXISTS HDEL HLEN HKEYS HVALS HGETALL
);

$cmd_type{$_} = 3 for qw(DEL MGET);   
$cmd_type{$_} = 4 for qw(MSET MSETNX);
$cmd_type{$_} = 5 for qw(BLPOP BRPOP);



sub readers {
	my ($c, $servers, $write2server, $write2client, $VERBOSE) = @_;

	my @cmd = ();
	my $cmd;     

	my $client_reader = get_client_reader(
		sub {
			my ($cmd_name, @args) = @_;
			print "REQUEST $cmd_name\n" if $VERBOSE;
			if (my $cmd_type = $cmd_type{$cmd_name}) {
				if ($cmd_type == 1) {
					push @cmd, [$cmd_name];
					my $buf = cmd2stream($cmd_name, @args);
					$write2server->($c, undef, $buf);
				} elsif ($cmd_type == 2) {
					my $s_addr = key2server($args[0], $servers);
					push @cmd, [$cmd_name, $s_addr];
					my $buf = cmd2stream($cmd_name, @args);
					$write2server->($c, $s_addr, $buf);
				} elsif ($cmd_type == 3) {
					my @s_addr = ();
					my %s_addr = ();
					for (@args) {
						my $s_addr = key2server($_, $servers);
						push @s_addr, $s_addr;
						push @{$s_addr{$s_addr}}, $_;
					}
					push @cmd, [$cmd_name, @s_addr];
					foreach my $s_addr (keys %s_addr) {
						my $buf = cmd2stream($cmd_name, @{$s_addr{$s_addr}});
						$write2server->($c, $s_addr, $buf);
					}
				} elsif ($cmd_type == 4) {
					my @s_addr = ();
					my %s_addr = ();
					my $i = 0;
					while ($i <= $#args) {
						my $key   = $args[$i++];
						my $value = $args[$i++];
						my $s_addr = key2server($key, $servers);
						push @s_addr, $s_addr;
						push @{$s_addr{$s_addr}}, $key, $value;
					}
					push @cmd, [$cmd_name, @s_addr];
					foreach my $s_addr (keys %s_addr) {
						my $buf = cmd2stream($cmd_name, @{$s_addr{$s_addr}});
						$write2server->($c, $s_addr, $buf);
					}
				} elsif ($cmd_type == 5) {
					my $timeout = $args[-1];
					my @s_addr = ();
					my %s_addr = ();
					for my $i (0 .. $#args - 1) {
						my $s_addr = key2server($args[$i], $servers);
						push @s_addr, $s_addr;
						push @{$s_addr{$s_addr}}, $args[$i];
					}
					if (keys %s_addr > 1) {
						$write2client->($c, "-ERR Keys of the '$cmd_name' command should be on one node; use key tags\015\012");
					} else {
						push @cmd, [$cmd_name, @s_addr];
						foreach my $s_addr (keys %s_addr) {
							my $buf = cmd2stream($cmd_name, @{$s_addr{$s_addr}}, $timeout);
							$write2server->($c, $s_addr, $buf);
						}
					}
				}
			} else {
				$write2client->($c, "-ERR unsupported command '$cmd_name'\015\012");
			}
		}
	);

	my $resp_type = "";
	my %resp_line      = ();
	my %resp_bulk_size = ();
	my %resp_bulk_args = ();
	my $size_all       = 0; 
	my $servers_reader = get_servers_reader(
		DEBUG   => 0,
		servers => $servers,
		sub_cmd => sub {
			if ($cmd = shift @cmd) {
				my ($cmd_name, @s_addr) = @$cmd;
				if (@s_addr) {
					my %s_addr = map { $_ => 1 } @s_addr;
					return $cmd_name, keys %s_addr;
				} else {
					return $cmd_name;
				}
			}
		},
		sub_response_type          => sub { ($resp_type) = @_ },
		sub_line_response          => sub { my ($s, $resp_line) = @_; $resp_line{$s} = $resp_line },
		sub_bulk_response_size     => sub { my ($s, $resp_bulk_size) = @_; $resp_bulk_size{$s} = $resp_bulk_size },
		sub_bulk_response_size_all => sub {
			$size_all = 1;
			if ($$cmd[0] eq "KEYS") {
				my $resp_bulk_size;
				$resp_bulk_size += $_ for grep { defined $_ } values %resp_bulk_size;
				if (defined $resp_bulk_size) {
				 	$write2client->($c, "*$resp_bulk_size\015\012");
				} else {
					$write2client->($c, "*-1\015\012");
				}
				if (keys %resp_bulk_args) {
					my $stream = args2stream(map { @$_ } values %resp_bulk_args);
					$write2client->($c, $stream);
				}
			}
		},
		sub_bulk_response_arg      => sub {
			my ($s, $arg) = @_;
			if ($size_all and $$cmd[0] eq "KEYS") {
				if (defined $arg) {
					$write2client->($c, join "", '$', length $arg, "\015\012", $arg, "\015\012");
				} else {
					$write2client->($c, "\$-1\015\012");
				}
			} else {
				push @{$resp_bulk_args{$s}}, $arg;
			}
		},
		sub_response_received      => sub {
			if ($resp_type eq "line") {
				my @v = values %resp_line;
				if (@v == 1) {
					$write2client->($c, "$v[0]\015\012");
				} else {
					my $v = shift @v;
					if ($v =~ m/^:\d+/ ) {
						my $sum = 0;
						$sum += $_ for map { m/^:(\d+)/ } $v, @v;
						$write2client->($c, ":$sum\015\012");
					} else {
						if (grep { $v ne $_ } @v) {
							$write2client->($c, "-ERR nodes return different results\015\012");
						} else {
							$write2client->($c, "$v\015\012");
						}
					}
				}
			} elsif ($resp_type eq "bulk") {

				if ($cmd_type{$$cmd[0]} != 2) {
					warn "bulk cmd $$cmd[0] with $cmd_type{$$cmd[0]} != 2";
				}

				if (keys %resp_bulk_size > 1) {
					warn "logic error";
				} else {
					my $s_addr = (keys %resp_bulk_size)[0];
					if (defined $resp_bulk_size{$s_addr}) {
						my $arg = $resp_bulk_args{$s_addr}[0];
						my $stream = join "", '$', length $arg, "\015\012", $arg, "\015\012";
						$write2client->($c, $stream);
					} else {
						$write2client->($c, "\$-1\015\012");
					}
				}

			} elsif ($resp_type eq "multi_bulk" and $$cmd[0] ne "KEYS") {
				my $resp_bulk_size;
				$resp_bulk_size += $_ for grep { defined $_ } values %resp_bulk_size;
				if (defined $resp_bulk_size) {
					my ($cmd_name, @s_addr) = @$cmd;
					my @args = ();
					if (@s_addr > 1) {
						foreach my $s_addr (@s_addr) {
							push @args, shift @{$resp_bulk_args{$s_addr}};
						}
					} else {
						push @args, map { @$_ } values %resp_bulk_args;
					}
				 	$write2client->($c, cmd2stream(@args));
				} else {
					$write2client->($c, "*-1\015\012");
				}
			}
			print "RESPONSE from all on $$cmd[0]\n" if $VERBOSE;
			$resp_type = "";
			%resp_line      = ();
			%resp_bulk_size = ();
			%resp_bulk_args = ();
			$size_all       = 0;
		},
	);

	return $client_reader, $servers_reader;
}



sub key2server {
	my ($key, $servers) = @_;
	$$servers[crc32($key =~ m/{(.+)}$/ ? $1 : $key) % @$servers];
}



sub cmd2stream {
	join "",
		'*', scalar @_, "\015\012",
		map {
			if (defined $_) {
				('$', length $_, "\015\012", $_, "\015\012");
			} else {
				('$-1', "\015\012");
			}
		} @_;
}

sub args2stream {
	join "",
		map {
			if (defined $_) {
				('$', length $_, "\015\012", $_, "\015\012");
			} else {
				('$-1', "\015\012");
			}
		} @_;
}


1;


resharding.pl
$| = 1;
use strict;
use warnings;

use RedisSharding;

use Getopt::Long;
use IO::Socket;
use EV;


GetOptions(
	"db=i"    => \ my $db,
	"from=s"  => \ my $from,
	"nodes=s" => \ my $nodes,
	"flushdb" => \ my $flushdb,
);

unless (defined $db and $from and $nodes) {
	print <<EOD;
Parameters 'db', 'from' and 'nodes' is required.

Using example:
perl $0 --db=9 --from=10.1.1.1:6379 --nodes=10.1.1.2:6380,10.1.1.3:6380,...
EOD
	exit;
}

my $blksize = 1024 * 16;

sub connecter {
	my ($host, $port) = split /:/, $_[0];
	IO::Socket::INET->new(Proto => 'tcp', Blocking => 0, PeerHost => $host, PeerPort => $port) or die $!;
}


my $fh_keys = connecter($from);
my $fh_type = connecter($from);

my @servers = split /\s*,\s*/, $nodes;
my %addr2fh = map { $_ => connecter($_) } @servers;

my %fh2rw = ();
my %fh2ww = ();


my %fh2buf = ();
my %fh2servers_reader = ();
my %fh2nodes_send_cmd = ();

my %fh2name = (
	$fh_keys => "fh_keys",
	$fh_type => "fh_type",
);
foreach (keys %addr2fh) {
	$fh2name{$addr2fh{$_}} = $_;
}

sub w_event_cb {
	my $w = shift;
	my $fh = $w->fh;
	my $len = syswrite $fh, $fh2buf{$fh}, $blksize;
	if ($len) {
		substr $fh2buf{$fh}, 0, $len, "";
		unless (length $fh2buf{$fh}) {
			delete $fh2ww{$fh};
		}
	}
}

sub r_event_cb {
	my $w = shift;
	my $fh = $w->fh;
	my $len = sysread $fh, (my $buf), $blksize;
	if ($len) {
		$fh2servers_reader{$fh}->($fh2name{$fh}, $buf);
	} elsif (defined $len) {
		delete $fh2rw{$fh};
		delete $fh2ww{$fh};
		close $fh;
		warn "ERROR: Redis server '$fh2name{$fh}' close connection";
		exit;
	}
};


my $progress = 0;
my $flag_all_keys;

my $counter = 0;
my $fh_keys_w_stoped = 0;
my $counter_keys = 0;
my $counter_type = 0;
my $counter_node_query = 0;
my $counter_node_resp  = 0;

sub counter_incr {
	$counter++;
	if (not $fh_keys_w_stoped and $counter > 100) {
		$fh_keys_w_stoped = 1;
		$fh2rw{$fh_keys}->stop();
	}
}
sub counter_decr {
	$counter--;
	if ($fh_keys_w_stoped and $counter < 50) {
		$fh_keys_w_stoped = 0;
		$fh2rw{$fh_keys}->start();
	}
	if ($flag_all_keys and $counter == 0) {
		unless (grep { $_ } map { length $_ } values %fh2buf) {
			print "\nDONE ($progress).\n";
			counter_show();
			exit;
		}
	}
}

sub counter_show {
	print "counter_keys $counter_keys\n";
	print "counter_type $counter_type\n";
	print "counter_node_query $counter_node_query\n";
	print "counter_node_resp  $counter_node_resp\n";
}


{
	my @cmd = ();
	my @allcmd = ();

	my $send_cmd = sub {
		if (not $fh_keys_w_stoped) {
		if (my $next_cmd = shift @allcmd) {
			$fh2buf{$fh_keys} .= cmd2stream(@$next_cmd);
			push @cmd, [$$next_cmd[0]];
			$fh2ww{$fh_keys} ||= EV::io($fh_keys, EV::WRITE, \&w_event_cb);
		}
		}
	};

	sub fh_keys_send_cmd {
		push @allcmd, \@_;
		$fh2buf{$fh_keys} or $send_cmd->();
	}

	my $cmd;
	$fh2servers_reader{$fh_keys} = get_servers_reader(
		servers => ["fh_keys"],
		sub_cmd                    => sub {
			if ($cmd = shift @cmd) {
				@$cmd;
			}
		},
		sub_response_type          => sub { my ($type) = @_ },
		sub_line_response          => sub { my ($s, $v) = @_;
			$v =~ m/^-ERR/ and die "ERROR on $$cmd[0]: ", $v;
		},
		sub_bulk_response_size     => sub { my ($s, $v) = @_;
			if ($$cmd[0] eq "KEYS") {
				print "DB $db contain $v keys.\n";
				unless ($v) {
					print "DONE.\n";
					exit;
				}
			} else {
				$v //= "-1"; print "RESPONSE from $s on $$cmd[0]; bulk size: $v\n"
			}
		},
		sub_bulk_response_size_all => sub { },
		sub_bulk_response_arg      => sub {
			my ($s, $v) = @_;
			if ($$cmd[0] eq "KEYS") {
				if (defined $v) {
					type_req($v);
					counter_incr();
					$counter_keys++;
				}
			} else {
				$v //= "-1";
				print "$s: $$cmd[0]: $v\n";
			}
		},
		sub_response_received      => sub {
			$send_cmd->();
			if ($$cmd[0] eq "KEYS") {
				$flag_all_keys = 1;
			}
		},
		DEBUG => 0,
	);

	$fh2rw{$fh_keys} = EV::io($fh_keys, EV::READ, \&r_event_cb);
}



{
	my @cmd = ();
	my @allcmd = ();

	my $send_cmd = sub {
		if (my $next_cmd = shift @allcmd) {
			$fh2buf{$fh_type} .= cmd2stream(@$next_cmd);
			push @cmd, [$$next_cmd[0]];
			$fh2ww{$fh_type} ||= EV::io($fh_type, EV::WRITE, \&w_event_cb);
		}
	};

	sub fh_type_send_cmd {
		push @allcmd, \@_;
		$fh2buf{$fh_type} or $send_cmd->();
	}

	my @k = ();
	sub type_req {
		my ($k) = @_;
		push @k, $k;
		fh_type_send_cmd("TYPE", $k);
	}

	my @kv = ();
	my @v  = ();

	my $DEBUG = 0;
	my $cmd;
	$fh2servers_reader{$fh_type} = get_servers_reader(
		servers => ["fh_type"],
		sub_cmd                    => sub {
			if ($cmd = shift @cmd) {
				@$cmd;
			}
		},
		sub_response_type          => sub { my ($type) = @_ },
		sub_line_response          => sub { my ($s, $v) = @_;
			$v =~ m/^-ERR/ and die "ERROR on $$cmd[0]: ", $v;
			if ($$cmd[0] eq "TYPE") {
				my $k = shift @k;
				print "$s: $$cmd[0]: $k is $v\n" if $DEBUG;
				push @kv, $k;
				if ($v eq "+string") {
					fh_type_send_cmd("GET", $k);
				} elsif ($v eq "+list") {
					fh_type_send_cmd("LRANGE", $k, 0, -1);
				} elsif ($v eq "+set") {
					fh_type_send_cmd("SMEMBERS", $k);
				} elsif ($v eq "+zset") {
					fh_type_send_cmd("ZRANGE", $k, 0, -1, "WITHSCORES");
				} elsif ($v eq "+hash") {
					fh_type_send_cmd("HGETALL", $k);
				} else {
					counter_decr();
					$counter_keys--;
				}
			}
		},
		sub_bulk_response_size     => sub { },
		sub_bulk_response_size_all => sub { },
		sub_bulk_response_arg      => sub { my ($s, $v) = @_; push @v, $v },
		sub_response_received      => sub {
			my %to_nodes = (
				GET => sub {
					my ($addr, $k, @v) = @_;
					fh_nodes_send_cmd($addr, "SET", $k, @v);
				},
				LRANGE => sub {
					my ($addr, $k, @v) = @_;
					foreach my $v (@v) {
						fh_nodes_send_cmd($addr, "RPUSH", $k, $v);
					}
				},
				SMEMBERS => sub {
					my ($addr, $k, @v) = @_;
					foreach my $v (@v) {
						fh_nodes_send_cmd($addr, "SADD", $k, $v);
					}
				},
				ZRANGE => sub {
					my ($addr, $k, @v) = @_; 
					my $i = 0;
					while ($i <= $#v) {
						my $v = $v[$i++];
						my $s = $v[$i++];
						fh_nodes_send_cmd($addr, "ZADD", $k, $s, $v);
					}
				},
				HGETALL => sub {
					my ($addr, $k, @v) = @_;
					fh_nodes_send_cmd($addr, "HMSET", $k, @v);
				},
			);
			if (my $to_nodes = $to_nodes{$$cmd[0]}) {
			 	my $k = shift @kv;
			 	my $addr = RedisSharding::key2server($k, \@servers);
				if ($DEBUG) {
					require Data::Dumper;
				 	print Data::Dumper::Dumper([$$cmd[0], $k, @v]);
				}
				$to_nodes->($addr, $k, @v) if @v;
				print "." unless ++$progress % 1000;
				counter_decr();
				$counter_type++;
			}
			@v = ();
			$send_cmd->();
		},
		DEBUG => 0,
	);

	$fh2rw{$fh_type} = EV::io($fh_type, EV::READ, \&r_event_cb);
}



foreach my $addr (keys %addr2fh) {
	my $fh = $addr2fh{$addr};

	my @cmd = ();
	my @allcmd = ();

	my $send_cmd = sub {
		if (my $next_cmd = shift @allcmd) {
			$fh2buf{$fh} .= cmd2stream(@$next_cmd);
			push @cmd, [$$next_cmd[0]];
			$fh2ww{$fh} ||= EV::io($fh, EV::WRITE, \&w_event_cb);
		}
	};

	$fh2nodes_send_cmd{$fh} = sub {
	 	push @allcmd, \@_;
	 	$fh2buf{$fh} or $send_cmd->();
	};

	my $cmd;
	$fh2servers_reader{$fh} = get_servers_reader(
		servers => [$addr],
		sub_cmd                    => sub {
			if ($cmd = shift @cmd) {
				@$cmd;
			}
		},
		sub_response_type          => sub { my ($type) = @_ },
		sub_line_response          => sub { my ($s, $v) = @_;
			$v =~ m/^-ERR/ and die "ERROR on $$cmd[0]: ", $v;
		},
		sub_bulk_response_size     => sub { },
		sub_bulk_response_size_all => sub { },
		sub_bulk_response_arg      => sub { },
		sub_response_received      => sub {
			$send_cmd->();
			$counter_node_resp++;
			counter_decr();
		},
		DEBUG => 0,
	);

	$fh2rw{$fh} = EV::io($fh, EV::READ, \&r_event_cb);
}

sub fh_nodes_send_cmd {
	my ($addr, @args) = @_;
	foreach my $fh ($addr ? $addr2fh{$addr} : values %addr2fh) {
		$fh2nodes_send_cmd{$fh}->(@args);
		counter_incr();
		$counter_node_query++;
	}
}


fh_keys_send_cmd("SELECT", $db);
fh_type_send_cmd("SELECT", $db);

fh_nodes_send_cmd(undef, "SELECT", $db);
fh_nodes_send_cmd(undef, "FLUSHDB") if $flushdb;

fh_keys_send_cmd("KEYS", "*");


EV::loop;
